File: services/crypto-service/src/vault.rs
use reqwest::Client;
use serde_json::Value;
use std::env;

#[derive(Clone)]
pub struct VaultClient {
    client: Client,
    vault_addr: String,
    vault_token: String,
}

impl VaultClient {
    pub fn new() -> Option<Self> {
        let vault_addr = env::var("VAULT_ADDR").unwrap_or_else(|_| "".to_string());
        let vault_token = env::var("VAULT_TOKEN").unwrap_or_else(|_| "".to_string());

        if vault_addr.is_empty() {
            log::warn!("VAULT_ADDR is not set, Vault integration is disabled.");
            return None;
        }

        Some(Self {
            client: Client::new(),
            vault_addr,
            vault_token,
        })
    }

    pub async fn get_system_keys(&self) -> Result<String, String> {
        let url = format!("{}/v1/secret/data/origin/crypto-service/system_keys", self.vault_addr);
        
        let response = self.client.get(&url)
            .header("X-Vault-Token", &self.vault_token)
            .send()
            .await
            .map_err(|e| format!("Network request to vault failed: {}", e))?;
            
        if !response.status().is_success() {
            return Err(format!("Vault returned error status: {}", response.status()));
        }
        
        let json: Value = response.json().await
            .map_err(|e| format!("Failed to parse vault response: {}", e))?;
            
        let key = json["data"]["data"]["escrow_agent_key"]
            .as_str()
            .ok_or_else(|| "Missing escrow_agent_key in Vault response".to_string())?;
            
        Ok(key.to_string())
    }
}

File: services/auth-service/core/vault.py
import os
import httpx
import logging

logger = logging.getLogger(__name__)

class VaultClient:
    def __init__(self):
        self.vault_addr = os.getenv("VAULT_ADDR", "")
        self.vault_token = os.getenv("VAULT_TOKEN", "")
        self.enabled = bool(self.vault_addr and self.vault_token)
        
        if not self.enabled:
            logger.warning("VAULT_ADDR or VAULT_TOKEN not set, Vault integration is disabled.")

    async def get_jwt_keys(self):
        """Fetch RSA keys for JWT signing from Vault."""
        if not self.enabled:
            return None, None
            
        url = f"{self.vault_addr}/v1/secret/data/origin/auth-service/jwt_keys"
        headers = {"X-Vault-Token": self.vault_token}
        
        try:
            async with httpx.AsyncClient() as client:
                resp = await client.get(url, headers=headers)
                resp.raise_for_status()
                data = resp.json()["data"]["data"]
                return data.get("private_key"), data.get("public_key")
        except Exception as e:
            logger.error(f"Failed to fetch keys from Vault: {e}")
            return None, None
